salt/loader.py

```diff
@@ -152,23 +134,93 @@ def _module_dirs(
             ext_type_dirs = "{}_dirs".format(tag)
         if ext_type_dirs in opts:
             ext_type_types.extend(opts[ext_type_dirs])
-        if HAS_PKG_RESOURCES and ext_type_dirs:
-            for entry_point in pkg_resources.iter_entry_points(
-                "salt.loader", ext_type_dirs
-            ):
-                try:
+        if ext_type_dirs:
+            for entry_point in entrypoints.iter_entry_points("salt.loader"):
+                with catch_entry_points_exception(entry_point) as ctx:
                     loaded_entry_point = entry_point.load()
-                    for path in loaded_entry_point():
+                if ctx.exception_caught:
+                    continue
+
+                # Old way of defining loader entry points
+                #   [options.entry_points]
+                #   salt.loader=
+                #     runner_dirs = thirpartypackage.loader:func_to_get_list_of_dirs
+                #     module_dirs = thirpartypackage.loader:func_to_get_list_of_dirs
+                #
+                #
+                # New way of defining entrypoints
+                #   [options.entry_points]
+                #   salt.loader=
+                #     <this-name-does-not-matter> = thirpartypackage
+                #     <this-name-does-not-matter> = thirpartypackage:callable
+                #
+                # We try and see if the thirpartypackage has a `ext_type` sub module, and if so,
+                # we append it to loaded_entry_point_paths.
+                # If the entry-point is in the form of `thirpartypackage:callable`, the return of that
+                # callable must be a dictionary where the keys are the `ext_type`'s and the values must be
+                # lists of paths.
```

Comment:
**suggestion**: not sure if we document it elsewhere but we should either mention that here or add some examples of the expected format for this.

Reply:
Yes, in an upcoming documentation only PR which walks you through the process of creating an extension from scratch